sys/net/application_layer/gcoap/gcoap.c

```diff
@@ -331,39 +330,41 @@ static ssize_t _finish_pdu(coap_pkt_t *pdu, uint8_t *buf, size_t len)
  * Finds the memo for an outstanding request within the _coap_state.open_reqs
  * array. Matches on token.
  *
- * src_pdu Source for the match token
+ * memo_ptr[out] -- Registered request memo, or NULL if not found
+ * src_pdu[in] -- PDU for token to match
  */
-static void _find_req_memo(gcoap_request_memo_t **memo_ptr, coap_pkt_t *src_pdu,
-                                                            uint8_t *buf, size_t len)
+static void _find_req_memo(gcoap_request_memo_t **memo_ptr, coap_pkt_t *src_pdu)
 {
-    gcoap_request_memo_t *memo;
-    coap_pkt_t memo_pdu = { .token = NULL };
-    (void) buf;
-    (void) len;
+    *memo_ptr = NULL;
+    /* no need to initialize struct; we only care about buffer contents below */
+    coap_pkt_t memo_pdu_data;
+    coap_pkt_t *memo_pdu = &memo_pdu_data;
+    unsigned cmplen      = coap_get_token_len(src_pdu);
 
     for (int i = 0; i < GCOAP_REQ_WAITING_MAX; i++) {
         if (_coap_state.open_reqs[i].state == GCOAP_MEMO_UNUSED)
             continue;
 
-        /* setup memo PDU from memo header */
-        memo                 = &_coap_state.open_reqs[i];
-        coap_hdr_t *memo_hdr = (coap_hdr_t *) &memo->hdr_buf[0];
-        memo_pdu.hdr         = memo_hdr;
-        if (coap_get_token_len(&memo_pdu)) {
-            memo_pdu.token = &memo_hdr->data[0];
+        gcoap_request_memo_t *memo = &_coap_state.open_reqs[i];
+        if (memo->send_limit == GCOAP_SEND_LIMIT_NON) {
+            memo_pdu->hdr = (coap_hdr_t *) &memo->msg.hdr_buf[0];
         }
-        /* match on token */
-        if (coap_get_token_len(src_pdu) == coap_get_token_len(&memo_pdu)) {
-            uint8_t *src_byte  = src_pdu->token;
-            uint8_t *memo_byte = memo_pdu.token;
-            size_t j;
-            for (j = 0; j < coap_get_token_len(src_pdu); j++) {
-                if (*src_byte++ != *memo_byte++) {
-                    break;      /* token mismatch */
+        else {
+            memo_pdu->hdr = (coap_hdr_t *) memo->msg.data.pdu_buf;
+        }
+
+        if (coap_get_token_len(memo_pdu) == cmplen) {
+            if (cmplen) {
```

Comment:
this can be optimised, i.e., by merging the 2 `if` to `if (cmplen && (coap_get_token_len(memo_pdu) == cmplen)) {`. However, I would rather put this into a helper function `_cmp_hdr_token(token A, token B)` which returns `-1`, `0`, or `1` like typical compare functions. That would make this a one line and this whole for-loop a lot more readable (IMHO).

Reply:
I just looked deeper into `memcmp()`, and I was wrong about the need to explicitly test when the comparison length is zero, as explained in this [SO question](https://stackoverflow.com/questions/19034858/return-value-of-memcmpptr1-ptr2-0).

So, the token comparison can be simplified to:
```
        if (coap_get_token_len(memo_pdu) == cmplen) {
            memo_pdu->token = &memo_pdu->hdr->data[0];
            if (memcmp(src_pdu->token, memo_pdu->token, cmplen) == 0) {
                *memo_ptr = memo;
                break;
            }
        }
```

Good enough?